### PR TITLE
RSDK 1020 Part I - refactor base, encoded motor struct and tests.

### PR DESCRIPTION
**Pull request by randhid**
Original PR: https://github.com/viamrobotics/rdk/pull/2211

### CODE REVIEW

File: components/motor/gpio/motor_encoder_test.go [hunk 1]
```diff
@@ -33,6 +35,7 @@ func (f *fakeDirectionAware) DirectionMoving() int64 {
 }
 
 func TestMotorEncoder1(t *testing.T) {
+	t.Skip()
```
Comment (randhid) [inline components/motor/gpio/motor_encoder_test.go:38]: @penguinland skips here will unblock you for Digital interrupts.

Comment (penguinland) [reply components/motor/gpio/motor_encoder_test.go:38]: My workaround to get https://github.com/viamrobotics/rdk/pull/2236 to pass the tests was to remove the one line that messed up the timing on the race conditions (https://github.com/viamrobotics/rdk/pull/2236/commits/b08b60d077676d8e882823d305d435306cc0650a, reverted in https://github.com/viamrobotics/rdk/pull/2236/commits/2c6a32af9c1caa6042e482a8868e2f69032420df). The rest is already merged. In the future, we can make the tests more robust, and add that log line back in.

File: components/motor/gpio/motor_encoder.go [hunk 2]
```diff
@@ -161,30 +158,36 @@ func newEncodedMotor(
 		em.flip = -1
 	}
 
+	if rdkutils.Float64AlmostEqual(em.ticksPerRotation, 0, 0.01) {
+		em.ticksPerRotation = 1
+	}
+
```
Comment (martha-johnston) [inline components/motor/gpio/motor_encoder.go:164]: [q] why does this need to use `Float64AlmostEqual` but `mc.TicksPerRotation` on line 64 can be directly compared to 0?

File: components/motor/gpio/motor_encoder.go [hunk 3]
```diff
@@ -399,17 +405,18 @@ func (m *EncodedMotor) rpmMonitorPass(pos, lastPos, now, lastTime int64, rpmDebu
 	// slow down so we don't overshoot
 	// halve and quarter rpm values based on seconds remaining in move
 
-	desiredRPM := m.state.desiredRPM
+	// desiredRPM := m.state.desiredRPM
```
Comment (martha-johnston) [inline components/motor/gpio/motor_encoder.go:408]: [nit] maybe remove unless you want to keep it for future reference since this is only part 1

**Review by martha-johnston — APPROVED** [review]
LGTM with the caveat that I don't 100% understand what some things are doing, but feel free to ignore my questions if you have a good understanding of the reasoning. I'll take one more look once the tests/checks are all passing

File: components/motor/gpio/motor_encoder.go (hunk 2 above)
Comment (randhid) [reply components/motor/gpio/motor_encoder.go:164]: good catch! thanks!

**Pull request merged**